### PR TITLE
Issue #3119 fix

### PR DESCRIPTION
**Pull request by saurabhjangir**
Fix and test cases added for Issue #3119

### CODE REVIEW

File: new [hunk 1]
```diff
@@ -0,0 +1 @@
+Just checking new changes
```
Comment (fgalan) [inline new]: Seems you git-add more than expected ;) This files should be removed

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# Copyright 2015 Telefonica Investigacion y Desarrollo, S.A.U
```
Comment (fgalan) [inline test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: 2015 -> 2018 ;)

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+HTTP/1.1 200 OK
+Content-Length: 128
+Content-Type: application/json
+Fiware-Correlator: REGEX([0-9a-f\-]{36})
+Date: REGEX(.*)
+
+{
+    "errorCode": {
+        "code": "400",
+        "details": "empty entityId:id",
```
Comment (fgalan) [inline test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: Details is misleading... in this case what we have is actually something like "null characters are not allowed in isPattern entity id" or a similar wordking.

File: src/lib/ngsi/EntityId.cpp [hunk 4]
```diff
@@ -140,7 +140,7 @@ std::string EntityId::toJson(void) const
 */
 std::string EntityId::check(RequestType requestType)
 {
-  if (id == "")
+  if (id == "" || ((id.find('\0')) != (std::string::npos)))
```
Comment (fgalan) [inline src/lib/ngsi/EntityId.cpp]: The C++ check here is maybe too cryptic an potencially ineficient in a field that is going to be checked very ofter (a large set of operations in the API use entities id). Mabye @kzangeli have some C-based alternative that could be evaluated.

File: CHANGES_NEXT_RELEASE [hunk 5]
```diff
@@ -3,3 +3,4 @@
 - Fix: case-sensitive header duplication (e.g. "Content-Type" and "Content-type") in custom notifications (#2893)
 - Fix: bug in GTE and LTE operations in query filters (q/mq), both for GET operations and subscriptions (#2995)
 - Fix: Wrong "max one service-path allowed for subscriptions" in NGSIv2 subscription operation (#2948)
+- Fix: Bug that may contextBroker to crash when passing inappropriate parameter in query context payload (#3055)
```
Comment (fgalan) [inline CHANGES_NEXT_RELEASE]: I think correct issue number is `#3119`.

In addition, given you are going to edit this CNR line, change contextBroker -> ContextBroker, pls.

**Comment by fgalan** [comment]
@saurabhjangir why you have closed this PR? Minor comments appart, it looks nice :)

**Comment by saurabhjangir** [comment]
I am sorry this request is pulled by mistake.
Please track this issue fix at #3125

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test (hunk 3 above)
Comment (saurabhjangir) [reply test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: Entity ID with Null characters are actually an empty ID.

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test (hunk 2 above)
Comment (saurabhjangir) [reply test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: ohk

File: src/lib/ngsi/EntityId.cpp (hunk 4 above)
Comment (saurabhjangir) [reply src/lib/ngsi/EntityId.cpp]: Yes we can do it. I did it for an immediate fix. As we need it ASAP.
I will look for the workaround soon :P

File: CHANGES_NEXT_RELEASE (hunk 5 above)
Comment (saurabhjangir) [reply CHANGES_NEXT_RELEASE]: Done. :)

Comment (saurabhjangir) [reply CHANGES_NEXT_RELEASE]: this was actually my last changed CNR ;)

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test (hunk 3 above)
Comment (fgalan) [reply test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: But you could have something like entity id = `\0afternull`. Strictly speaking that is not empty ID, but it has null... or not?

File: src/lib/ngsi/EntityId.cpp (hunk 4 above)
Comment (kzangeli) [reply src/lib/ngsi/EntityId.cpp]: Yeah, this is kind of strange ... '\0' is simply 0, there is no difference.
And the check is to make sure there is no ZERO inside the string. There cannot be any zero inside a string in C. The zero is the marker that the string has ended ... Very cryptid indeedd. I understand nothing of this.  The functest tells me it works though ... Might be something weird about std::string but ... beats me.

What happens if you do this...?:

```
std::string s = "abcde";
char*   cP = (char*) s.c_str();

cP[2] = 0;
```

In C ... something similar would set `s` to "ab". Position 2 is now the end of the string ...
Looking forward to hearing an explanation of all this ... :-)

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test (hunk 3 above)
Comment (kzangeli) [reply test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: As `"\0"` is 0, the string should end there and it is empty ...

File: src/lib/ngsi/EntityId.cpp (hunk 4 above)
Comment (kzangeli) [reply src/lib/ngsi/EntityId.cpp]: Ah!  Could it have something to do with UTF?  "Funny" charsets ... ?

File: test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test (hunk 3 above)
Comment (saurabhjangir) [reply test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: @fgalan Your concern was fixed now. Previously we were getting empty ID error messages. After few changes which I've just commited please find the request and response below:
```
[root@localhost fiware-orion]# (curl localhost:1026/v1/queryContext -s -S --header 'Content-Type: application/json' \
>     --header 'Accept: application/json' -d @- | python -mjson.tool) <<EOF
> {
>     "entities": [
>         {
>             "type": "Room",
>             "isPattern": "true",
>             "id": "\u0000"
>         }
>     ]
> }
> EOF
{
    "errorCode": {
        "code": "400",
        "details": "invalid regex for entity id pattern",
        "reasonPhrase": "Bad Request"
    }
}
```

Comment (saurabhjangir) [reply test/functionalTest/cases/3119_qc_nullcharacter_payload/qc_nullcharacter_payload.test]: Please track this at PR #3125

**Pull request closed without merging**